doc: Added doc for running ingress/GwAPI in host network mode (#31839)
cilium-team/cilium: Documentation/network/servicemesh/gateway-api/host-network-mode.rst

.. only:: not (epub or latex or html)

    WARNING: You are looking at unreleased Cilium documentation.
    Please use the official rendered version released here:
    https://docs.cilium.io

Host network mode
#################

.. note::
    Supported since Cilium 1.16+

> This mode allows you to expose the Cilium Gateway API Gateway directly on the

Reviewer: Nit, let's decouple the section title with the text:

```suggestion
Host network mode mode allows you to expose the Cilium Gateway API Gateway directly on the
```

host network.

> This is useful in some edge cases where one doesn't want, or can't, use
> LoadBalancer type services (dev environments, cluster-external loadbalancer,
> ...).

Reviewer: Two nits here:
1. Let's remove the parenthesis, as these are good examples to state directly in the sentence.
2. This sentence uses "one" to refer to the reader, but the previous sentence uses "you".

```suggestion
This is useful in edge cases where a LoadBalancer Service is unavailable, such
as in development environments or environments with cluster-external loadbalancers.
```

.. note::
    * Enabling the Cilium Gateway API host network mode automatically disables the LoadBalancer type service mode. They are mutually exclusive.
    * The listener is exposed on all interfaces (``0.0.0.0`` for IPv4 and/or ``::`` for IPv6).

> The host network mode can be enabled via Helm flag:

Reviewer: Grammar nit:

```suggestion
Host network mode can be enabled via Helm:
```

.. code-block:: yaml

    gatewayAPI:
      enabled: true
      hostNetwork:
        enabled: true

> Once enabled, the host network ports can be specified on the ``Gateway``
> resources via ``spec.listeners.port``. The port must be unique per ``Gateway``

Reviewer: Is there only one host network port per Gateway resource? If so, can we reword to make this more clear?

```suggestion
Once enabled, the host network port for a ``Gateway`` can be specified via ``spec.listeners.port``.
The port must be unique per ``Gateway``
```

> resource and you should choose a port number higher than 1023 (see

Reviewer: Nit: in the next section, the port number is formatted as a code sample, so let's do the same here:

```suggestion
resource and you should choose a port number higher than ``1023`` (see
```

`Bind to privileged port`_).

.. warning::
    Be aware that misconfiguration might result in port clashes. Configure unique ports that are still available on all Cilium Nodes where Gateway API listeners are exposed.

Bind to privileged port
=======================

> 
> By default, it's mandatory to use ports higher than ``1023`` as the Cilium L7
> Envoy process does not have any Linux capabilities out-of-the-box (so no
> ``CAP_NET_BIND_SERVICE``) and is therefore not allowed to listen on privileged
> ports.
> If you choose a port equal to or lower than ``1023``, ensure that
> ``envoy.securityContext.capabilities.keepCapNetBindService=true`` is configured
> and to add the capability ``NET_BIND_SERVICE`` to the respective
> :ref:`Cilium Envoy container via Helm values<envoy>`:

Reviewer: `By default` and `it's mandatory` are a little conflicting, and may confuse folks who are skimming. Can we reword to make it clear that a configuration change is required for binding to privileged ports?

```suggestion

By default, the Cilium L7 Envoy process does not have any Linux capabilities
out-of-the-box and is therefore not allowed to listen on privileged ports.

If you choose a port equal to or lower than ``1023``, ensure that the helm value
``envoy.securityContext.capabilities.keepCapNetBindService=true`` is configured
and to add the capability ``NET_BIND_SERVICE`` to the respective
:ref:`Cilium Envoy container via Helm values<envoy>`:
```

* Standalone DaemonSet mode: ``envoy.securityContext.capabilities.envoy``
* Embedded mode: ``securityContext.capabilities.ciliumAgent``

> Configure the following to allow privileged port bindings in host network mode:

Reviewer: Nit: let's make it clear that these are helm values

```suggestion
Configure the following helm values to allow privileged port bindings in host network mode:
```

> .. code-block:: yaml

Reviewer: Could we split this code block up into tabs, where each tab represents the required change for each type of envoy mode? I think it would be easier to read.

    gatewayAPI:
      enabled: true
      hostNetwork:
        enabled: true
    # For Cilium Envoy standalone DaemonSet mode:
    envoy:
      enabled: true
      securityContext:
        capabilities:
          keepCapNetBindService: true
          envoy:
          # Add NET_BIND_SERVICE to the list (keep the others!)
          - NET_BIND_SERVICE
    # For Cilium Envoy embedded mode:
    envoy:
      securityContext:
        capabilities:
          keepCapNetBindService: true
    securityContext:
      capabilities:
        ciliumAgent:
        # Add NET_BIND_SERVICE to the list (keep the others!)
        - NET_BIND_SERVICE

Deploy Gateway API listeners on subset of nodes
===============================================

> It's supported to expose the Cilium Gateway API Envoy listener only on specific
> nodes. This only works in combination with the host network mode and can be
> configured via a node label selector in the Helm values.

Reviewer: Couple of nits:

1. Use a colon to transition to the helm values.
2. In the first sentence, stating that the functionality "is supported" is a bit redundant, since we can just say what is supported outright:

```suggestion
The Cilium Gateway API Envoy listener can be exposed on a specific subset of
nodes. This only works in combination with the host network mode and can be
configured via a node label selector in the Helm values:
```

.. only:: not (epub or latex or html)

    WARNING: You are looking at unreleased Cilium documentation.
    Please use the official rendered version released here:
    https://docs.cilium.io

Host network mode
#################

.. note::
    Supported since Cilium 1.16+

Host network mode allows you to expose the Cilium Gateway API Gateway directly
on the host network.

> This is useful in edge cases where a LoadBalancer Service is unavailable, such

Reviewer 2: nit: rather than saying this is edge cases, can we just mention that there may be some cases where this is needed?

as in development environments or environments with cluster-external
loadbalancers.

.. note::

>     * Enabling the Cilium Gateway API host network mode automatically disables the LoadBalancer type service mode. They are mutually exclusive.

Reviewer 2: nit: capitalize Service to indicate we're talking about a k8s service.

Author: Thanks for those two inputs, Tom. I addressed them in https://github.com/cilium/cilium/pull/31839/commits/2be1399fe7c56d0d44112f6bb5502668cb4d705c.

Host network mode
#################

> 

Reviewer 3: ```suggestion
.. warning::
    Supported since Cilium 1.16+
```

> Since Cilium 1.16+, it's supported to expose the Cilium GatewayAPI Gateway

Reviewer 3: ```suggestion
This mode allows you to expose the Cilium GatewayAPI Gateway
```

Host network mode
#################
.. note::
    Supported since Cilium 1.16+

This mode allows you to expose the Cilium GatewayAPI Gateway
directly on the host network.
This is useful in some edge cases where one don't want or can't to use
LoadBalancer type services (dev environments, cluster-external
loadbalancer, ...).

The host network mode can be enabled via Helm flag
``gatewayAPI.hostNetwork.enabled=true``).
Once enabled, the host network ports can be specified on the ``Gateway``
resources via ``spec.listeners.port``. The port must be unique per ``Gateway``
resource and higher than ``1023``.

.. warning::
    * The listener is exposed on all interfaces (``0.0.0.0`` for IPv4 and/or ``::`` for IPv6).
    * Be aware that misconfiguration might result in port clashes. Configure unique ports which are still available.
    * It's mandatory to use ports higher than ``1023`` as the Cilium L7 proxy process `does not have any Linux capabilities`_ (so no ``CAP_NET_BIND_SERVICE``) and is therefore not allowed to listen on privileged ports.

> .. _does not have any Linux capabilities: https://github.com/cilium/proxy/pull/315

Reviewer 4: FYI: this PR alone doesn't solve the issue. in addition we need to 

- add the capability to the containers (proxy (daemonset mode) and agent (embedded mode))
- pass the new flag to the envoy starter

Author: Right, that is a valid point. I can also link to the other to-be-created PRs once they exist.

Reviewer 4: PRs are open
- changing default port for shared and dedicated ingress from `80` to unprivileged `8080`: https://github.com/cilium/cilium/pull/32159
- Support binding to privileged ports: https://github.com/cilium/cilium/pull/32158

Thanks Philip!

Furthermore, it's supported to expose the Cilium GatewayAPI Envoy
listener only on specific nodes. This only works in combination with the
host network mode and can be configured via a node label selector in the Helm
values. Here is a complete host network mode example configuration for
Cilium GatewayAPI:

.. code-block:: yaml

    gatewayAPI:
      enabled: true
      hostNetwork:
        enabled: true
        nodes:
          matchLabels:
            role: infra

>             component: ingress

Reviewer 4: `component: gateway-api` in case of Gateway API?

Author: Thanks, fixed via https://github.com/cilium/cilium/pull/31839/commits/dd3a425a33e98731c25168c6122dda22403165f0.

Host network mode
#################
.. note::
    Supported since Cilium 1.16+

This mode allows you to expose the Cilium GatewayAPI Gateway
directly on the host network.

> This is useful in some edge cases where one don't want or can't to use

Reviewer 5: ```suggestion
This is useful in some edge cases where one doesn't want, or can't, use
```

LoadBalancer type services (dev environments, cluster-external
loadbalancer, ...).

The host network mode can be enabled via Helm flag

> ``gatewayAPI.hostNetwork.enabled=true``).

Reviewer 5: ```suggestion
``gatewayAPI.hostNetwork.enabled=true``.
```

Once enabled, the host network ports can be specified on the ``Gateway``
resources via ``spec.listeners.port``. The port must be unique per ``Gateway``

> resource and higher than ``1023``.

Reviewer 4: I like the wording on the Ingress page better:

> you should choose a port number higher than ``1023`` (see warning below).

Maybe we can adapt this and "weaken" the `must` (which is more targeting the `unique`ness of the port

> .. warning::

Reviewer 4: I think this is more of a `note` than a `warning` - except the following

> Be aware that misconfiguration might result in port clashes. Configure unique ports that are still available.

Maybe worth to change into two blocks (note and warning)?

    * Enabling the Cilium GatewayAPI host network mode automatically disables the LoadBalancer/NodePort type service mode. They are mutually exclusive.
    * The listener is exposed on all interfaces (``0.0.0.0`` for IPv4 and/or ``::`` for IPv6).

>     * Be aware that misconfiguration might result in port clashes. Configure unique ports that are still available.

Reviewer 4: > Configure unique ports that are still available.

... that are still available on all Cilium Nodes where Gateway API listeners are exposed.

>     * Be aware that misconfiguration might result in port clashes. Configure unique ports which are still available.

Reviewer 5: ```suggestion
    * Be aware that misconfiguration might result in port clashes. Configure unique ports that are still available.
```

>     * It's mandatory to use ports higher than ``1023`` as the Cilium L7 proxy process `does not have any Linux capabilities by default`_ (so no ``CAP_NET_BIND_SERVICE``) and is therefore not allowed to listen on privileged ports. If you choose a port equals to or lower than ``1023``, ensure to configure ``envoy.securityContext.capabilities.keepCapNetBindService=true`` and to add the capability ``NET_BIND_SERVICE`` to the respective :ref:`Cilium Envoy container via Helm values<envoy>`:

Reviewer 5: ```suggestion
    * It's mandatory to use ports higher than ``1023`` as the Cilium L7 proxy process `does not have any Linux capabilities by default`_ (so no ``CAP_NET_BIND_SERVICE``) and is therefore not allowed to listen on privileged ports. If you choose a port equal to or lower than ``1023``, ensure that ``envoy.securityContext.capabilities.keepCapNetBindService=true`` is configured and to add the capability ``NET_BIND_SERVICE`` to the respective :ref:`Cilium Envoy container via Helm values<envoy>`:
```

>     * It's mandatory to use ports higher than ``1023`` as the Cilium L7 proxy process `does not have any Linux capabilities by default`_ (so no ``CAP_NET_BIND_SERVICE``) and is therefore not allowed to listen on privileged ports. If you choose a port equal to or lower than ``1023``, ensure that ``envoy.securityContext.capabilities.keepCapNetBindService=true`` is configured and to add the capability ``NET_BIND_SERVICE`` to the respective :ref:`Cilium Envoy container via Helm values<envoy>`:

Reviewer 4: `Cilium L7 proxy process` -> `Cilium L7 Envoy process` ?

        * Standalone DaemonSet mode: ``envoy.securityContext.capabilities.envoy``
        * Embedded mode: ``securityContext.capabilities.ciliumAgent``

> .. _does not have any Linux capabilities by default: https://github.com/cilium/proxy/pull/315

Reviewer 4: is this still required? it seems as this doesn't get rendered at all https://deploy-preview-31839--docs-cilium-io.netlify.app/network/servicemesh/gateway-api/gateway-api#host-network-mode

Author: Not really. Let's remove it.

Furthermore, it's supported to expose the Cilium GatewayAPI Envoy
listener only on specific nodes. This only works in combination with the
host network mode and can be configured via a node label selector in the Helm
values. Here is a complete host network mode example configuration for
Cilium GatewayAPI:

.. code-block:: yaml

    gatewayAPI:
      enabled: true
      hostNetwork:
        enabled: true
        nodes:
          matchLabels:
            role: infra
            component: gateway-api
    # -- Optional: Allow privileged port bindings (not recommended)
    #
    # Configure the following to allow privileged port bindings in host network mode.
    #
    # For Cilium Envoy standalone DaemonSet mode:

>     envoy:

Reviewer 4: i wonder whether it would be worth to split this example up into `node label selector` and `privileged port` - maybe even with dedicated sub-headings

Author: 💯, done.

      enabled: true
      securityContext:
        capabilities:
          keepCapNetBindService: true
          envoy:
          # Add NET_BIND_SERVICE to the list (keep the others!)
          - NET_BIND_SERVICE
    # For Cilium Envoy embedded mode:
    envoy:
      securityContext:
        capabilities:
          keepCapNetBindService: true
    securityContext:
      capabilities:
        ciliumAgent:
        # Add NET_BIND_SERVICE to the list (keep the others!)
        - NET_BIND_SERVICE

> An empty selector selects all nodes and continues to expose the functionality

Reviewer 4: Maybe prepend the sentence:

> This will deploy the Gateway API Envoy listener only on the Cilium Nodes matching the configured labels.

Followed by `An empty selector ...`

Author: Done.
